daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -28,156 +28,224 @@ import org.xml.sax.Locator
+ *
+ * 1. A DaffodilUnparseContentHandler instance is created, which initializes a
+ *    SAXInfosetInputter instance.
+ *
+ * 2. An XMLReader instance is created and configured to use the
+ *    DaffodilUnparseContentHandler to handle its SAX events.
+ *
+ * 3. The DaffodilUnparseContentHandler handles events from the XMLReader, gathers
+ *    the necessary information from those events, and fills out an array of
+ *    SAXInfosetEvent objects, called a "batch".
+ *
+ * 4. When a full batch has been gathered, or an endDocument() event is handled, the
+ *    DaffodilUnparseContentHandler calls resume() to pause execution and start the
+ *    SAXInfosetInputter coroutine, sending it the batch of events.
+ *
+ * 5. Since this is the first time the SAXInfosetInputter has been given control, its
+ *    coroutine Thread is started and its run() function called. This calls
+ *    waitResume() to receive the first batch of events. It then calls the
+ *    DataProcessor unparse() function to begin the unparse.
+ *
+ * 6. The SAXInfosetInputter functions are called by the unparse, which reads the
+ *    batched events and provides the necessary information to unparse the infoset.
+ *    Once the SAXInfosetInputter has unparsed all batched events and needs a new
+ *    event in the next() function, it calls resume() to pause execution and resume
+ *    the DaffodilUnparseContentHandler coroutine, sending it Nope to signify it
+ *    needs more events.
+ *
+ * 7. The DaffodilUnparseContentHandler resumes and again continues to handle SAX
+ *    events and fill out the batched events array until it is full or endDocument()
+ *    is handled, at which point it calls resume() to pause and send the new batch of
+ *    events to the SAXInfosetInputter where it resumes control.
+ *
+ * 8. Steps 6 and 7 repeat until the SAXInfosetInputter signals back to the
+ *    DaffodilUnparseContentHandler that it is complete calling resumeFinal() and
+ *    sending a One object containing either an UnparseResult or an Exception. At
+ *    this point, the SAXInfosetInputter is complete and provide control back to the
+ *    DaffodilUnparseContentHandler--the SAXInfosetInputter coroutine is done.
+ *
+ * 9. The DaffodilUnparseContentHandler resumes, recieves and examines the result
+ *    from the SAXInfosetInputter, and either makes the UnparseResult available to
+ *    the XMLReader or throws a SAXException if there was an error.
+ *
+ * @param dp DataProcessor object that will be used to start the unparse
+ * @param output OutputChannel of where the unparsed data is written
  */
 class DaffodilUnparseContentHandler(
   dp: DFDL.DataProcessor,
   output: DFDL.Output)
-  extends DFDL.DaffodilUnparseContentHandler {
-  private lazy val inputter = new SAXInfosetInputter(this, dp, output)
-  private var unparseResult: DFDL.UnparseResult = _
+  extends MainCoroutine[Maybe[Either[Exception, DFDL.UnparseResult]]]
+  with DFDL.DaffodilUnparseContentHandler {
```

Comment:
Just making another observation.  This line changes DaffodilUnparseContentHandler from MainCoroutine[Array[SAXInfosetEvent]] to MainCoroutine[Maybe[Either[Exception, DFDL.UnparseResult]]].  SAXInfosetInputter is still Coroutine[Array[SAXInfosetEvent]].  What this means is that a pair of coroutines can send data to each other without the data having to be the same type on both ends:

```scala
    // dataToReceive and dataToSend can be different types!
    val dataToReceive = this.resume(otherCoroutine, dataToSend)
```

This is valuable to know - maybe worth adding to a comment somewhere.

Reply:
Yeah, I didn't know this either until I started digging more deeply into this SAX an coroutine stuff. I'll make sure it's clear somewhere (maybe the Coroutine.scala documentation) that the type parameter is the type the coroutine expects to receive from its peer when it calls resume, which are allowed to be different.